src/Acuminator/Acuminator.Analyzers/StaticAnalysis/EventHandlerModifier/EventHandlerModifierAnalyzer.cs

```diff
@@ -0,0 +1,121 @@
+﻿using System.Collections.Immutable;
+using Acuminator.Analyzers.StaticAnalysis.PXGraph;
+using Acuminator.Utilities.Roslyn.Semantic.PXGraph;
+using Acuminator.Utilities.Roslyn.Semantic;
+using Microsoft.CodeAnalysis;
+using System.Linq;
+using Microsoft.CodeAnalysis.Diagnostics;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.CSharp;
+
+namespace Acuminator.Analyzers.StaticAnalysis.PrivateEventHandlers
+{
+	public class EventHandlerModifierAnalyzer : PXGraphAggregatedAnalyzerBase
+	{
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+			ImmutableArray.Create(
+				Descriptors.PX1077_EventHandlersShouldBeProtectedVirtual,
+				Descriptors.PX1077_EventHandlersShouldNotBeExplicitInterfaceImplementations);
+		
+		private static readonly Dictionary<SyntaxKind, string> ModifierText = new()
+		{
+			{ SyntaxKind.PublicKeyword, "public" },
+			{ SyntaxKind.ProtectedKeyword, "protected" },
+			{ SyntaxKind.VirtualKeyword, "virtual" }
+		};
+
+		private static readonly Dictionary<Accessibility, SyntaxKind> AccessibilitySyntaxKindMap = new()
+		{
+			{ Accessibility.Public, SyntaxKind.PublicKeyword },
+			{ Accessibility.Protected, SyntaxKind.ProtectedKeyword }
+		};
+
+		public override void Analyze(SymbolAnalysisContext context, PXContext pxContext, PXGraphEventSemanticModel graphOrGraphExtension)
+		{
+			context.CancellationToken.ThrowIfCancellationRequested();
+			
+			var declaredEventHandlers = graphOrGraphExtension
+				.GetAllEvents()
+				.Where(graphEvent => graphEvent.Symbol.IsDeclaredInType(graphOrGraphExtension.Symbol));
+
+			var allInterfaces = graphOrGraphExtension
+				.Symbol
+				.AllInterfaces
+				.SelectMany(im => im.GetMethods())
+				.Where(im => im.GetEventHandlerType(pxContext) != EventType.None)
+				.ToList();
+
+			foreach (var handler in declaredEventHandlers)
+			{
+				context.CancellationToken.ThrowIfCancellationRequested();
+
+				var location = handler.Symbol.Locations.FirstOrDefault();
+
+				if (location == null)
+				{
+					continue;
+				}
+
+				if (handler.Symbol.IsOverride)
+				{
+					continue;
+				}
+
+				var properties = new Dictionary<string, string?>
+				{
+					{ DiagnosticProperty.IsContainingTypeSealed, handler.Symbol.ContainingType.IsSealed.ToString() }
```

Comment:
After reading the implementation after the fixes I have understood that the case of an `abstract` event handler was not supported. We can't add `virtual` keyword to an abstract event handler, that will be a compiler error.

I think, a good way to handle this is to create another `DiagnosticProperty` - "AddVirtual" flag, for example.
You can set it like this:
```cs
bool addVirtualModifier = !handler.Symbol.ContainingType.IsSealed && !handler.Symbol.IsAbstract;
```
Then pass it to the code fix in properties.

Reply:
Done.